### ISSUE TITLE
MooseMesh::isSemiLocal() returns false negatives

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#9869]
### Description of the enhancement or error report

Querying ```MooseMesh::isSemiLocal(node)``` appears to return true iff ```node``` is either on a local element or on an *explicitly*-ghosted element.  It returns false if ```node``` is on neither of those things but is on an *implicitly* ghosted element, in which case we have a false negative.

### Rationale for the enhancement or information for reproducing the error

```
METHOD=dbg ./run_tests -p 2 --re 'phase_field/test:MultiSmoothCircleIC.multi_normal_test'
```

Triggers the problem for me, when the false negative causes an assert to fail.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

In the short term, I can replace any affected uses of ```isSemiLocal()``` with more specific (albeit less terse and convenient) tests.  However, ```getActiveSemiLocalNodeRange()``` is affected by the same bug, and at this point I wouldn't be surprised if users of that method were *depending* on the current semantics, so I don't think it's safe to change the implementation here.

In the long term I think I'd prefer to deprecate these two methods entirely and introduce something with a new name, which would have the correct (and nothing but the correct) algebraic-semilocal nodes from the start regardless of ghosting settings.

**Comment by permcody** [comment]
You have my blessing to clean up any of this junk. I'm also open to handing off support to libMesh. The genesis of these methods is important to understand. Some of these methods are old (definitely predating the ghosting functor work) and we didn't have the support we needed in libMesh. We could either just deprecate some of these methods and start using the newer methods in libMesh or point these existing methods to methods in libMesh. We've had a lot of misunderstandings over the years of what semi-local and ghosted meant. I know we used those interchangeably in MOOSE and that has caused problems.

**Comment by roystgnr** [comment]
Yeah, I probably could have helped avert the whole mess by creating better documentation for the "semilocal" terminology, by which I mean *some* documentation for the "semilocal" terminology.

On the other hand, it really wasn't until I had to hammer down the ghosting functor APIs that we even had a solid set of definitions to document.  If I'd had to pin myself down earlier I might have missed key distinctions (e.g. that an object even on a distributed mesh might need to be "geometrically" semilocal despite not being "algebraically" semilocal) and then we'd just have *that* update as a backwards compatibility issue to worry about.

**Comment by roystgnr** [comment]
On the gripping hand, there may not even need to be a way to test whether something is geometrically semilocal or not, because if you can even get your hands on a pointer to it then the answer is "yes", and if you're not sure whether you can get your hands on a pointer to it or not then that's what query_elem_ptr and query_node_ptr are for.  Admittedly, those don't distinguish between "this id is of a remote object" and "this id is of a nonexistent object", but in general if you even have an id to look up then you already know it refers to something which exists.

**Comment by permcody** [comment]
> Yeah, I probably could have helped avert the whole mess by creating better documentation for the "semilocal" terminology, by which I mean some documentation for the "semilocal" terminology.

Cool - So that means when we wrote this (or Derek probably), we were "not even wrong"!
